nixos/modules/services/web-apps/booklore.nix

```diff
@@ -0,0 +1,330 @@
+        EnvironmentFile = cfg.environmentFile;
+
+        # Hardening
+        CapabilityBoundingSet = "";
+        NoNewPrivileges = true;
+        PrivateUsers = true;
+        PrivateTmp = true;
+        PrivateDevices = true;
+        PrivateMounts = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        RestrictAddressFamilies = [
+          "AF_INET"
+          "AF_INET6"
+          "AF_UNIX"
+        ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+      };
+    };
+
+    services.booklore.environment = {
+      BOOKLORE_DATA_DIR = cfg.dataDir;
+      BOOKLORE_BOOKDROP_DIR = cfg.bookdropDir;
+      DATABASE_HOST = if cfg.database.host != null then cfg.database.host else "127.0.0.1";
+      DATABASE_PORT = if cfg.database.port != null then toString cfg.database.port else "3306";
+      DATABASE_USERNAME = cfg.database.user;
+      SERVER_PORT = toString cfg.api.port;
+    };
+
+    services.nginx = lib.mkIf cfg.nginx.enable {
+      enable = true;
+      virtualHosts.${cfg.nginx.virtualHost} = {
+        listen = [
+          {
+            addr = cfg.host;
+            port = cfg.port;
+          }
+        ];
+
+        forceSSL = cfg.nginx.forceSSL;
+        enableACME = lib.mkDefault (cfg.nginx.forceSSL && cfg.nginx.useACMEHost == null);
+        useACMEHost = cfg.nginx.useACMEHost;
+
+        root = "${cfg.package}/share/booklore-ui";
+
+        extraConfig = ''
+          client_max_body_size ${cfg.settings.maxBodySize};
+        '';
+
+        locations = {
+          "/" = {
+            tryFiles = "$uri $uri/ /index.html =404";
+            extraConfig = ''
+              location ~* \.mjs$ {
+                types {
+                    text/javascript mjs;
+                }
+              }
+            '';
+          };
+
+          "/api/" = {
+            proxyPass = "http://localhost:${toString cfg.api.port}";
+            recommendedProxySettings = lib.mkDefault true;
+          };
+
+          "/ws" = {
+            proxyPass = "http://localhost:${toString cfg.api.port}/ws";
+            recommendedProxySettings = lib.mkDefault true;
+            extraConfig = ''
+              proxy_http_version 1.1;  # Ensure HTTP 1.1 is used for WebSocket connection
+              proxy_set_header Upgrade $http_upgrade;  # Pass the upgrade header
+              proxy_set_header Connection 'upgrade';  # Pass the connection header
+            '';
+          };
+        };
+      };
+    };
+
+    services.mysql = lib.mkIf cfg.database.createLocally {
+      enable = true;
+
+      settings = {
+        bind_address = lib.mkIf (cfg.database.host != null) cfg.database.host;
+        port = lib.mkIf (cfg.database.port != null) cfg.database.port;
+      };
```

Comment:
I really don't think it's a good idea for the `booklore` to decide what socket `mysql` should bind to. I would never expect this as a downstream user. If anything, `booklore` could default to connect to the address decided by `services.mysql` if `createLocally` is enabled.